src/pip/_internal/utils/misc.py

```diff
@@ -480,6 +481,40 @@ def user_test(d):
             ]
 
 
+def search_distribution(req_name):
+
+    # Canonicalize the name before searching in the list of
+    # installed distributions and also while creating the package
+    # dictionary to get the Distribution object
+    req_name = canonicalize_name(req_name)
+    packages = get_installed_distributions(skip=())
+    pkg_dict = {canonicalize_name(p.key): p for p in packages}
+    return pkg_dict.get(req_name)
+
+
+def get_distribution(req_name):
+    """Given a requirement name, return the installed Distribution object"""
+
+    # Search the distribution by looking through the working set
+    dist = search_distribution(req_name)
+
+    # If distribution could not be found, call working_set.require
+    # to update the working set, and try to find the distribution
+    # again This might happen for e.g. when you install a package
```

Comment:
```suggestion
    # again. This might happen for e.g. when you install a package
```

How would this case happen though? Doesn’t pip uninstall the editable package when you install the non-editable one (or vice versa)?

Reply:
To be more clear, I was referring to the [test_uninstall.test_uninstall_setuptools_develop_install](https://github.com/pypa/pip/blob/master/tests/functional/test_uninstall.py#L506) which was also observed in https://github.com/pypa/pip/pull/8054#discussion_r425405545  

Are editable and non-editable installs different then using `setup.py install` and `setup.py develop` respectively? If yes, I can change the language accordingly.

So what I observed was that after adding the `working_set.require` change, the logs after running the commands as above are as follows.

<details>
<summary>logs with the change</summary>

```
$ python setup.py develop
/Users/devesh/pip/.env/lib/python3.8/site-packages/setuptools/dist.py:471: UserWarning: Normalizing '0.1dev' to '0.1.dev0'
  warnings.warn(
running develop
running egg_info
writing FSPkg.egg-info/PKG-INFO
writing dependency_links to FSPkg.egg-info/dependency_links.txt
writing entry points to FSPkg.egg-info/entry_points.txt
writing top-level names to FSPkg.egg-info/top_level.txt
reading manifest file 'FSPkg.egg-info/SOURCES.txt'
writing manifest file 'FSPkg.egg-info/SOURCES.txt'
running build_ext
Creating /Users/devesh/pip/.env/lib/python3.8/site-packages/FSPkg.egg-link (link to .)
Adding FSPkg 0.1.dev0 to easy-install.pth file

Installed /Users/devesh/pip/tests/data/packages/FSPkg
Processing dependencies for FSPkg==0.1.dev0
Finished processing dependencies for FSPkg==0.1.dev0

$ python setup.py install
/Users/devesh/pip/.env/lib/python3.8/site-packages/setuptools/dist.py:471: UserWarning: Normalizing '0.1dev' to '0.1.dev0'
  warnings.warn(
running install
running bdist_egg
running egg_info
writing FSPkg.egg-info/PKG-INFO
writing dependency_links to FSPkg.egg-info/dependency_links.txt
writing entry points to FSPkg.egg-info/entry_points.txt
writing top-level names to FSPkg.egg-info/top_level.txt
reading manifest file 'FSPkg.egg-info/SOURCES.txt'
writing manifest file 'FSPkg.egg-info/SOURCES.txt'
installing library code to build/bdist.macosx-10.15-x86_64/egg
running install_lib
running build_py
creating build/bdist.macosx-10.15-x86_64/egg
creating build/bdist.macosx-10.15-x86_64/egg/fspkg
copying build/lib/fspkg/__init__.py -> build/bdist.macosx-10.15-x86_64/egg/fspkg
byte-compiling build/bdist.macosx-10.15-x86_64/egg/fspkg/__init__.py to __init__.cpython-38.pyc
creating build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/PKG-INFO -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/SOURCES.txt -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/dependency_links.txt -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/entry_points.txt -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/not-zip-safe -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/top_level.txt -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
creating 'dist/FSPkg-0.1.dev0-py3.8.egg' and adding 'build/bdist.macosx-10.15-x86_64/egg' to it
removing 'build/bdist.macosx-10.15-x86_64/egg' (and everything under it)
Processing FSPkg-0.1.dev0-py3.8.egg
creating /Users/devesh/pip/.env/lib/python3.8/site-packages/FSPkg-0.1.dev0-py3.8.egg
Extracting FSPkg-0.1.dev0-py3.8.egg to /Users/devesh/pip/.env/lib/python3.8/site-packages
Removing FSPkg 0.1.dev0 from easy-install.pth file
Adding FSPkg 0.1.dev0 to easy-install.pth file

Installed /Users/devesh/pip/.env/lib/python3.8/site-packages/FSPkg-0.1.dev0-py3.8.egg
Processing dependencies for FSPkg==0.1.dev0
Finished processing dependencies for FSPkg==0.1.dev0

$ pip list --format=json
[{"name": "appdirs", "version": "1.4.4"}, {"name": "distlib", "version": "0.3.0"}, {"name": "filelock", "version": "3.0.12"}, {"name": "FSPkg", "version": "0.1.dev0"}, {"name": "mypy", "version": "0.770"}, {"name": "mypy-extensions", "version": "0.4.3"}, {"name": "packaging", "version": "20.3"}, {"name": "pip", "version": "20.2.dev0"}, {"name": "pluggy", "version": "0.13.1"}, {"name": "py", "version": "1.8.1"}, {"name": "pyparsing", "version": "2.4.7"}, {"name": "setuptools", "version": "41.2.0"}, {"name": "six", "version": "1.14.0"}, {"name": "toml", "version": "0.10.1"}, {"name": "tox", "version": "3.15.0"}, {"name": "typed-ast", "version": "1.4.1"}, {"name": "typing-extensions", "version": "3.7.4.2"}, {"name": "virtualenv", "version": "20.0.20"}]

$ pip uninstall FSPKG -y
Found existing installation: FSPkg 0.1.dev0
Uninstalling FSPkg-0.1.dev0:
  Successfully uninstalled FSPkg-0.1.dev0

$ pip uninstall FSPKG -y
Found existing installation: FSPkg 0.1.dev0
Uninstalling FSPkg-0.1.dev0:
  Successfully uninstalled FSPkg-0.1.dev0
```

</details>

As you can see both installations go through, and then calling the uninstallation twice works as well. But once I remove the `working_set.require` change, the output is as follows

<details>
<summary>logs without the change</summary>

```
$ python setup.py develop
/Users/devesh/pip/.env/lib/python3.8/site-packages/setuptools/dist.py:471: UserWarning: Normalizing '0.1dev' to '0.1.dev0'
  warnings.warn(
running develop
running egg_info
writing FSPkg.egg-info/PKG-INFO
writing dependency_links to FSPkg.egg-info/dependency_links.txt
writing entry points to FSPkg.egg-info/entry_points.txt
writing top-level names to FSPkg.egg-info/top_level.txt
reading manifest file 'FSPkg.egg-info/SOURCES.txt'
writing manifest file 'FSPkg.egg-info/SOURCES.txt'
running build_ext
Creating /Users/devesh/pip/.env/lib/python3.8/site-packages/FSPkg.egg-link (link to .)
Adding FSPkg 0.1.dev0 to easy-install.pth file

Installed /Users/devesh/pip/tests/data/packages/FSPkg
Processing dependencies for FSPkg==0.1.dev0
Finished processing dependencies for FSPkg==0.1.dev0

$ python setup.py install
/Users/devesh/pip/.env/lib/python3.8/site-packages/setuptools/dist.py:471: UserWarning: Normalizing '0.1dev' to '0.1.dev0'
  warnings.warn(
running install
running bdist_egg
running egg_info
writing FSPkg.egg-info/PKG-INFO
writing dependency_links to FSPkg.egg-info/dependency_links.txt
writing entry points to FSPkg.egg-info/entry_points.txt
writing top-level names to FSPkg.egg-info/top_level.txt
reading manifest file 'FSPkg.egg-info/SOURCES.txt'
writing manifest file 'FSPkg.egg-info/SOURCES.txt'
installing library code to build/bdist.macosx-10.15-x86_64/egg
running install_lib
running build_py
creating build/bdist.macosx-10.15-x86_64/egg
creating build/bdist.macosx-10.15-x86_64/egg/fspkg
copying build/lib/fspkg/__init__.py -> build/bdist.macosx-10.15-x86_64/egg/fspkg
byte-compiling build/bdist.macosx-10.15-x86_64/egg/fspkg/__init__.py to __init__.cpython-38.pyc
creating build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/PKG-INFO -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/SOURCES.txt -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/dependency_links.txt -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/entry_points.txt -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/not-zip-safe -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
copying FSPkg.egg-info/top_level.txt -> build/bdist.macosx-10.15-x86_64/egg/EGG-INFO
creating 'dist/FSPkg-0.1.dev0-py3.8.egg' and adding 'build/bdist.macosx-10.15-x86_64/egg' to it
removing 'build/bdist.macosx-10.15-x86_64/egg' (and everything under it)
Processing FSPkg-0.1.dev0-py3.8.egg
creating /Users/devesh/pip/.env/lib/python3.8/site-packages/FSPkg-0.1.dev0-py3.8.egg
Extracting FSPkg-0.1.dev0-py3.8.egg to /Users/devesh/pip/.env/lib/python3.8/site-packages
Removing FSPkg 0.1.dev0 from easy-install.pth file
Adding FSPkg 0.1.dev0 to easy-install.pth file

Installed /Users/devesh/pip/.env/lib/python3.8/site-packages/FSPkg-0.1.dev0-py3.8.egg
Processing dependencies for FSPkg==0.1.dev0
Finished processing dependencies for FSPkg==0.1.dev0

$ pip list --format=json
[{"name": "appdirs", "version": "1.4.4"}, {"name": "distlib", "version": "0.3.0"}, {"name": "filelock", "version": "3.0.12"}, {"name": "FSPkg", "version": "0.1.dev0"}, {"name": "mypy", "version": "0.770"}, {"name": "mypy-extensions", "version": "0.4.3"}, {"name": "packaging", "version": "20.3"}, {"name": "pip", "version": "20.2.dev0"}, {"name": "pluggy", "version": "0.13.1"}, {"name": "py", "version": "1.8.1"}, {"name": "pyparsing", "version": "2.4.7"}, {"name": "setuptools", "version": "41.2.0"}, {"name": "six", "version": "1.14.0"}, {"name": "toml", "version": "0.10.1"}, {"name": "tox", "version": "3.15.0"}, {"name": "typed-ast", "version": "1.4.1"}, {"name": "typing-extensions", "version": "3.7.4.2"}, {"name": "virtualenv", "version": "20.0.20"}]

$ pip uninstall FSPKG -y
Found existing installation: FSPkg 0.1.dev0
Uninstalling FSPkg-0.1.dev0:
  Successfully uninstalled FSPkg-0.1.dev0

$ pip uninstall FSPKG -y
WARNING: Skipping FSPKG as it is not installed.
```
</details>

As you can see, the second uninstallation fails here. 

Based on these observations, I added the comment. I think the comment can be re-worded to better explain what's happening here, so any suggestions to change that is welcome.

Reply:
I see, so the problem is when the package is *not* installed by pip; `setup.py` does not uninstall packages because, well, it don’t know how. Yeah, it would be better to clarify the comment (because `pip install` won’t have the problem).

Reply:
@deveshks Could you add this clarification?

Reply:
I updated the comment according to the clarification asked for.